Add "required" task option

goldenapples opened the pull request:
Adds a setting field for "required" tasks which can be set when
registering tasks. If a task is registered with "required" => true, then
that task's callback must return true in order for a post to be
transitioned into publish or future status.

If a user attempts to publish a post which does not meet all required
tasks' requirements, the post will remain in its previous status and a
message will be displayed in admin notices alerting the user of the
condition(s) which must be met in order to publish the post.

"Required" tasks are also displayed first in the checklist, in bold text
for emphasis.

goldenapples commented:
This is how the warning appears:
 
<img width="1264" alt="screen shot 2016-02-17 at 11 41 33 am" src="https://cloud.githubusercontent.com/assets/665992/13116833/71b75ee6-d56b-11e5-8994-ff9a65ae558b.png">

The biggest problem here that I see is that an unfamiliar user could easily ignore this as yet another warning notice, and not realize that their post didn't successfully publish.

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -97,12 +100,71 @@ public function action_post_submitbox_misc_actions_render_checklist() {
+	 * @param string $old_status
+	 * @param object $post
+	 */
+	public function action_transition_post_status( $new_status, $old_status, $post ) {
```
For a variety of reasons, I think forcefully reverting the status is a pretty bad idea, with the potential for unintended consequences.

goldenapples replied:
Hm. And there's the conundrum, I guess. 

The checklist tasks are going to be written in such a way as to operate on the post itself, rather than on a $_POST array which would be accessible before the post was updated. 

The only other way I can see of reasonably blocking publishing is to hook in early enough on save_post to change the status to draft before saving, then set it back to the original status again and run the transition post status hooks again after all checks have passed. I'll give that a shot.

danielbachhuber replied:
> The only other way I can see of reasonably blocking publishing is to hook in early enough on save_post to change the status to draft before saving, then set it back to the original status again and run the transition post status hooks again after all checks have passed. I'll give that a shot.

Or you could just hide the publish button. Blocking the saving process is bad UX and will likely cause bugs.

goldenapples replied:
> Or you could just hide the publish button.

Makes sense.

goldenapples replied:
Trying something different in 28a6c52 - removing the "Publish" capability if a post doesn't have all required checklist items completed. Not sure if I like it or not yet. In theory, the "Submit for Review" status is nice to use for this, but I'm not sure if its too confusing in practice.

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -35,6 +35,7 @@ public static function get_instance() {
 	 * Set up actions for the plugin
 	 */
 	private function setup_actions() {
+		add_filter( 'user_has_cap', array( $this, 'filter_user_has_cap' ), 10, 4 );
```
I think `map_meta_cap` is the technically-correct way of adjusting user permissions on the fly. However, I'm not sure whether there are any practical differences.

goldenapples commented:
Updated in 28a6c52 with a different kind of UI:

Now, posts which have incomplete "required" checklist items will not have a Publish button available to them. Instead, the progress bar will have a notice about the required tasks which have to be met in order to publish, and the major publishing action available will be "Submit for Review":

<img width="272" alt="screen shot 2016-02-17 at 3 17 23 pm" src="https://cloud.githubusercontent.com/assets/665992/13123536/958a21e0-d58a-11e5-9aa7-fe5d738947a0.png">

On completion of all required tasks, the "publish" capability will be available again:

<img width="270" alt="screen shot 2016-02-17 at 3 18 15 pm" src="https://cloud.githubusercontent.com/assets/665992/13123569/b8c8c242-d58a-11e5-9c6d-31f2ae0ed35a.png">

The code is pretty messy at this point and could be cleaned up and refactored a bunch before shipping. But I'd appreciate any feedback on the UI. @danielbachhuber, @ryanboswell?

danielbachhuber commented:
> Instead, the progress bar will have a notice about the required tasks which have to be met in order to publish, and the major publishing action available will be "Submit for Review":

I think this works. Can you / should you be able to save as draft though?

goldenapples commented:
> Can you / should you be able to save as draft though?

Oh, yeah. You can, and should be able to. In my screenshot, I had already submitted an incomplete post once, so it was showing as "Pending". Until a post has been "submitted for review", the "Save as Draft" button is available:

<img width="274" alt="screen shot 2016-02-17 at 3 46 34 pm" src="https://cloud.githubusercontent.com/assets/665992/13124164/e05f8158-d58d-11e5-821d-3130cec569aa.png">

Its somewhat strange that there's no way for a non-super admin to switch a pending post back to draft status. But that's kinda just how core is.

Another weird UI issue to work around is that, since the capability check is done on rendering the page, if all checklist items are updated and I save the draft, I'll see a Publish button. But then I can undo one of the the required checklist items and publish, getting a nonsensical message telling me to complete some tasks before coming back and submitting my post for review - even though it's already been published.

<img width="274" alt="screen shot 2016-02-17 at 3 53 10 pm" src="https://cloud.githubusercontent.com/assets/665992/13124463/9cc4d41a-d58e-11e5-8190-f01a54b07199.png">

Though... I suppose it'd be easy enough to just render a different message if the post is already published.

goldenapples commented:
Updated and cleaned up. Code review, anyone?

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -55,16 +56,42 @@ private function setup_actions() {
 
+		if ( $publish_cap === $cap ) {
+			$required_tasks = $this->evaluate_checklist( $post->ID );
+			if ( $required_tasks['incomplete_required'] ) {
```
`evaluate_checklist()` returns `false` if the post doesn't exist, so we'll want to make sure to avoid an error notice here.

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -55,16 +56,42 @@ private function setup_actions() {
+	 * Exceptions are superadmins, who can do anything, and users with the
+	 * capability `publishing_checklist_publish_incomplete_posts`.
+	 */
+	public function filter_map_meta_cap( $all_caps, $cap ) {
```
This should be `$all_caps, $cap, $user_id, $args`

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -55,16 +56,42 @@ private function setup_actions() {
+		if ( $publish_cap === $cap ) {
+			$required_tasks = $this->evaluate_checklist( $post->ID );
+			if ( $required_tasks['incomplete_required'] ) {
+				$all_caps[ $publish_cap ] = current_user_can( 'publishing_checklist_publish_incomplete_posts' );
```
You should use `user_can()` on the `$user_id` passed through to be more precise

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -55,16 +56,42 @@ private function setup_actions() {
+	 * capability `publishing_checklist_publish_incomplete_posts`.
+	 */
+	public function filter_map_meta_cap( $all_caps, $cap ) {
+		global $post;
```
You should use `$args[0]` passed through above to be more precise.

danielbachhuber commented on publishing-checklist.php:
```diff
@@ -55,16 +56,42 @@ private function setup_actions() {
+		$publish_cap = $post_type_object->cap->publish_posts;
 
+		if ( $publish_cap === $cap ) {
+			$required_tasks = $this->evaluate_checklist( $post->ID );
```
What are the performance considerations of running this a hundred times on a page load? Would it be better to have an `evaluate_required_checklist()` or similar method to only evaluate required tasks, to ensure we aren't running tasks unnecessarily?

danielbachhuber commented:
Looking better — left some comments.

> Until a post has been "submitted for review", the "Save as Draft" button is available:

I wonder how difficult it would be to modify the button text in this situation.

r-a-y commented:
> I wonder how difficult it would be to modify the button text in this situation, so we could use a more correct label.

Hi, we're testing out this PR for a project and we agree that the "Submit for Review" label should be changed to something like "Check Requirements".

We're currently doing this by switching out the label via JS:

```
add_action( 'post_submitbox_misc_actions', function( $post ) {
	// This should dynamically use PC's post type list
	if ( 'POST TYPE TO CHECK' !== $post->post_type ) {
		return;
	}

	$js = <<<JS

<script>
jQuery(function($){
	if ( $('.publishing-checklist progress').val() !== $('.publishing-checklist progress').prop('max') ) {
		$('#publish').val('Check Requirements');
	}
})
</script>

JS;

	echo $js;
} );
```

The alternative is to use WP's `'gettext'` filter, which can be overkill.

dpjustice commented:
Anything I can do to help get this through? :)
